src/nccl_ofi_rdma.c

```diff
@@ -5328,15 +5314,27 @@ static int connect(nccl_net_ofi_ep_t *base_ep,
 
 		comm_state->stage = COMM_RECV_CONN;
 
-	case COMM_RECV_CONN:
+	case COMM_RECV_CONN: {
 		/* COMM_RECV_CONN: Receive connect response message from remote */
 
-		ret = post_recv_conn_resp(s_comm, device, ep);
-		if (ret == -FI_EAGAIN) {
-			return 0;
-		} else if (ret != 0) {
-			send_close(s_comm);
-			return ret;
+		bool recv_conn_resp_posted = false;
+		while (!recv_conn_resp_posted) {
+			ret = post_recv_conn_resp(s_comm, device, ep);
+			if (ret == -FI_EAGAIN) {
+				/* Block until we post the connection response request.
+				EAGAIN only involves waiting for local resources to free up, so it
+				should be safe to block. */
+				ret = ofi_process_cq(ep);
+				if (OFI_UNLIKELY(ret != 0)) {
+					send_close(s_comm);
+					return ret;
+				}
+			} else if (ret != 0) {
+				send_close(s_comm);
```

Comment:
I don't entirely understand why we're closing the send comm here?

Reply:
It's just freeing up resources in case of error. Although we might be freeing up memory that is in use by Libfabric or another thread, so I'm not sure this is the right thing to do either.

Reply:
this is one of those places you need to be sure.

Reply:
Ack, I'm pretty sure it's incorrect to close `s_comm` after sending the connect message, without waiting for that to complete. We do it in a couple other places in this function (from before this PR), but I won't touch those here. But I'll remove the close calls from this PR.